### ISSUE TITLE
Mongodb 1.2.4 driver strange behavior with 'readPreference' => 'nearest'

### ISSUE THREAD

**Issue by sautchenkov** [mongodb/mongo-php-driver#522]
We have 3 MongoDB nodes with replication settings. Now we using 1.2.0-alpha3 mongodb driver and want to update to stable. We waited for [this bug fix](https://github.com/mongodb/mongo-php-driver/issues/516) and today try to use fresh 1.2.4 mongodb driver but received strange behavior.

For fast read we using 'readPreference' => 'nearest'
In 1.2.0-alpha3 this working good and always read from local mongo. But as we see in 1.2.4 this read from any random node. 

Test script:
```
<?php

error_reporting(E_ALL);
ini_set('display_errors', 'On');
$start_time = microtime(true);

include __DIR__ . '/../../admin/components/Autoload.php';

$preference = NULL;
$limit = 100;
$sort = ['id' => -1];

if( isset($_GET['primary']) ) {
	$preference = 'primary';
}
elseif( isset($_GET['nearest']) ) {
	$preference = 'nearest';
}
elseif( isset($_GET['secondary']) ) {
	$preference = 'secondary';
}

$user = 'some_user';
$password = 'some_password';

if($preference == NULL) {
	$hosts = 'localhost:27017';
	$client = new \MongoDB\Client('mongodb://'. $user .':'. $password .'@'. $hosts);
}
else {
	$hosts = 'hetzner-1:27017,belarus-1:27017,selectel-1:27017';
	$client = new \MongoDB\Client(
		'mongodb://'. $user .':'. $password .'@'. $hosts,
		[
			'connectTimeoutMS' => 2000,
			'replicaSet' => 'rs0',
			'w' => 1,
			'wtimeoutMS' => 3000,
			'journal' => 'false',
			'readConcernLevel' => 'local',
			'readPreference' => $preference
		],
		[
			'typeMap' => ['root' => 'array', 'document' => 'array', 'array' => 'array']
		]
	);
}

print_r($client->getManager()->getReadPreference()); echo '<br>';

$collection = $client->selectCollection('domozon_reshop_by', 'customers');

$cursor = $collection->find([], ['sort' => $sort, 'limit' => $limit]);

print_r($cursor->getServer()->getHost());

$i = 1;
foreach ($cursor as $document) {
	if($i > 10 && $i < ($limit - 2)) {
		$i++;
		continue;
	}
	echo '<pre>';
	echo $i++ . ' ' . $document['_id'] . ' '. $document['id'] .'<br>';
	echo '</pre>';
}

$end_time = microtime(true);
echo 'Started '. $start_time.'<br>';
echo 'Ended '. $end_time.'<br>';
exit('Executed in '. round($end_time - $start_time, 4));

```
In $cursor->getServer()->getHost() we receive randome mongodb node in each request. In 1.2.0-alpha3 this is always local mongodb hostname. 

Software:
MongoDB 3.2.11
PHP 7.0.10
MongoDB driver: 1.2.4 stable installed from pecl
MongoDB library: 1.1.1

**Comment by jmikola** [comment]
The behavior you're seeing in the latest driver version is correct. Per the [Server Selection Algorithm](https://docs.mongodb.com/manual/core/read-preference-mechanics/) article in the MongoDB manual, `nearest` is evaluated like so:

> * The driver assembles a list of eligible members (primary and secondaries). maxStalenessSeconds and tag sets can further limit the eligibility of the members.
> * If the list of eligible members is not empty, the driver determines which eligible member is the “closest” (i.e. the member with the lowest average network round-trip-time) and calculates a latency window by adding the average round-trip-time of this “closest” server and the localThresholdMS [1]. The driver uses this latency window to pare down the list of eligible members to those members that fall within this window.
> * From this list of eligible members that fall within the latency window, the driver randomly chooses an eligible member.

For additional reference, this is also discussed in the server selection specification: [nearest](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#nearest) and [selecting servers within the latency window](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#selecting-servers-within-the-latency-window).

The behavior you were seeing in 1.2.0alpha3 was actually a bug in libmongoc, whereby "nearest" only selected the server with the lowest latency instead of a random server within the allowed latency window. Our 1.2.0 release bumped the libmongoc dependency to 1.5.0, which included [CDRIVER-1639](https://jira.mongodb.org/browse/CDRIVER-1639) and brought its server selection logic up to spec.

If you'd actually like to have "nearest" always select the server with the lowest latency, you can specify zero for the [localThresholdMS](https://docs.mongodb.com/manual/reference/connection-string/#urioption.localThresholdMS) URI option (either in the connection string or second array parameter to the `MongoDB\Client` or `MongoDB\Driver\Manager` constructor).

**Comment by sautchenkov** [comment]
Thanks, @jmikola  for the additional explanation about "nearest" algorithm. 
But as we tested:
1. All nodes exclude local have more than 25 ms RTT
selectel-1  RTT 0 ms    local node

hetzner-1  RTT 48 ms 
belarus-1  RTT 21 ms
backup-1  RTT 62 ms

As I understand with default localThresholdMS (15 ms) only selectel-1 (local node) need to be select as nearest.

2. We try to set localThresholdMS to 0 but behavior not changed.

You can look updated test script:

```
<?php

error_reporting(E_ALL);
ini_set('display_errors', 'On');
$start_time = microtime(true);

include __DIR__ . '/../../admin/components/Autoload.php';

$limit = 100;
$sort = ['id' => -1];
$preference = NULL;

if( isset($_GET['primary']) ) {
	$preference = 'primary';
}
elseif( isset($_GET['nearest']) ) {
	$preference = 'nearest';
}
elseif( isset($_GET['secondary']) ) {
	$preference = 'secondary';
}

$user = 'some_user';
$password = 'some_password';

if($preference == NULL) {
	$hosts = 'localhost:27017';
	$client = new \MongoDB\Client('mongodb://'. $user .':'. $password .'@'. $hosts);
}
else {
	$hosts = 'hetzner-1:27017,belarus-:27017,selectel-1:27017';
	$client = new \MongoDB\Client(
		'mongodb://'. $user .':'. $password .'@'. $hosts,
		[
			'connectTimeoutMS' => 2000,
			'replicaSet' => 'rs0',
			'w' => 1,
			'wtimeoutMS' => 3000,
			'journal' => 'false',
			'readConcernLevel' => 'local',
			'readPreference' => $preference,
			'localThresholdMS' => 0
		],
		[
			'typeMap' => ['root' => 'array', 'document' => 'array', 'array' => 'array']
		]
	);
}

print_r($client->getManager()->getReadPreference()); echo '<br>';

$collection = $client->selectCollection('domozon_reshop_by', 'customers');

$cursor = $collection->find([], ['sort' => $sort, 'limit' => $limit]);

echo 'Select '; print_r($cursor->getServer()->getHost());

echo '<pre>';
print_r($client->getManager()->getServers());
echo '</pre>';

$i = 1;
foreach ($cursor as $document) {
	if($i > 10 && $i < ($limit - 2)) {
		$i++;
		continue;
	}
	echo '<pre>';
	echo $i++ . ' ' . $document['_id'] . ' '. $document['id'] .'<br>';
	echo '</pre>';
}

$end_time = microtime(true);
echo 'Started '. $start_time.'<br>';
echo 'Ended '. $end_time.'<br>';
exit('Executed in '. round($end_time - $start_time, 4));
```
You can look work of this script in [this url this](http://reserv.recoms.net/test/mongo?nearest)

**Comment by sautchenkov** [comment]
Everybody who using stable mongodb 1.2.x+ replication + nearest receive this error or just only we?  :) 

We still can't upgrade to stable mongodb driver... Maybe we need to sent additional information or write to another place?

**Comment by jmikola** [comment]
> everybody who using stable mongodb 1.2.x+ replication + nearest receives this behavior or just only we? :) This is bug or we did something wrong?

After some more investigation, this appears to be a bug related to libmongoc's parsing of URI options. Zero is considered an unset value and libmongoc falls back to its default of 15ms when fetching the "localThresholdMS" option. I've opened [CDRIVER-2056](https://jira.mongodb.org/browse/CDRIVER-2056) to track the issue. One workaround would be to use `1` instead of `0`. The driver would still make a random selection if multiple servers were within 1ms of the nearest measured latency, but I believe this will work just fine for your topology, as the non-local servers all have much greater latency differentials (according to the [hosted script](http://reserv.recoms.net/test/mongo?nearest) you shared).

> We still can't upgrade to stable mongodb driver... Maybe we need to sent additional information or write to another place?

Just to confirm, are you referring only to this "localThresholdMS" issue or are there other outstanding issues preventing you from upgrading?

**Comment by sautchenkov** [comment]
@jmikola  Thanks! set localThresholdMS 1 ms resolve our trouble and this starts working as we need. 

But I want to pay attention. We think bug not only in interpret 0 as default 15 ms. We don't have server nearest 20 ms and if we not set localThresholdMS (driver need use the default 15 ms) mongo read from a random server.
If we set localThresholdMS as 15 ms it's working as need and read only from local. It seems like default localThresholdMS not 15 ms. At least this strange what not set localThresholdMS and set localThresholdMS as 15 ms is different.

**Comment by jmikola** [comment]
> It seems like default localThresholdMS not 15 ms.

I can confirm that is a separate bug in libmongoc. I've opened [CDRIVER-2057](https://jira.mongodb.org/browse/CDRIVER-2057) to track it and submitted https://github.com/mongodb/mongo-c-driver/pull/423 with a fix. I'm not sure when that fix will be released, or if it will end up in 1.5.x or only 1.6.x, so I've also opened [PHPC-917](https://jira.mongodb.org/browse/PHPC-917) to address it in the PHP driver (we can assign the correct default ourselves when constructing the URI).

Thanks for the detailed follow-up and helping uncover these issues!

**Comment by jmikola** [comment]
For reference, I've also opened [PHPC-918](https://jira.mongodb.org/browse/PHPC-918) as a local tracking ticket for the original bug with parsing zero ([CDRIVER-2056](https://jira.mongodb.org/browse/CDRIVER-2056)).

**Comment by jmikola** [comment]
If you're using bundled versions of libmongoc/libbson (default for PECL installs), this should be fixed in the [1.2.6](https://github.com/mongodb/mongo-php-driver/releases/tag/1.2.6) version of the extension. If you're using libmongoc as a shared library, upgrading it to [1.5.5](https://github.com/mongodb/mongo-c-driver/releases/tag/1.5.5) or [1.6.1](https://github.com/mongodb/mongo-c-driver/releases/tag/1.6.1) should resolve things.
